### PR TITLE
Tmyrden/14 template

### LINKED ISSUE #14: Allow Custom Templating
We currently allow the developer to specify a container to put the user list within. However, it would be nice to allow the developer to specify a template to use for the user list.

When pairing user list with the use of JWTs, the extra headers you can add for user details aren't able to be rendered into the user list.

For example, I added an extra header to the JWT that contains the user's email. In order to make their name in the user list a mailto link, I have to use jQuery to modify the DOM after the user list is rendered. Would be a nice feature-add to give the developer the option to customize the template used inside the user list.

### PR DESCRIPTION
**Pull request by tmyrden**
Closes #14 

Adds custom templating to the user view.

### CODE REVIEW

File: docs/v1/user_list.md [hunk 1]
```diff
@@ -171,11 +171,12 @@ Creates the UserList instance with customizable options.
...
-| - `avatars` [**default true**] is a [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean) that enables or disables user avatars.
-| - `userOptions` [**default true**] is a [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean) that enables or disables user options, which allows users to change their display name.
+| - `avatars` [**default: true**] is a [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean) that enables or disables user avatars. |
+| - `userOptions` [**default: true**] is a [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean) that enables or disables user options, which allows users to change their display name. |
+| - `userTemplate` [**default: null**] is a [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String) of the HTML that will be used to render the rows of the user list. The default template can be found [here.](https://github.com/goinstant/user-list/blob/master/templates/user-template.html) |
```
Comment (nealstewart) [inline docs/v1/user_list.md]: Ee. I don't know how I feel about linking to the GitHub repo directly.

File: docs/v1/user_list.md [hunk 2]
```diff
@@ -190,6 +191,46 @@ var options = {
...
+
+| Template Variables |
+|:---|
+| Type: [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) |
+| The user template is passed an object with the following properties: |
+| - `user` is the [User Object](../javascript_api/users.html). The user object also contains any extraHeaders specified inside a [JWT](../security_and_auth/guides/index.html).|
+| - `shortName` is truncated version of the user displayName.|
+| - `avatarColor` [**default: 'right'**]is a [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String) that specifies the user color, if the [User Colors](./user_colors.html) widget is in use. |
+| - `avatarUrl` is a [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String) that gives the URL of the user's avatar. |
+| - `loaded` is a [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean) that indicates if the user's avatar image has loaded. |
```
Comment (nealstewart) [inline docs/v1/user_list.md]: was able to be loaded

File: lib/user_view.js [hunk 3]
```diff
@@ -78,7 +80,13 @@ UserView.prototype.render = function(user, cb) {
       tmplVars.loaded = true;
     }
 
-    var template = _.template(userTemplate, tmplVars);
+    var template;
+    if (self._userTemplate) {
```
Comment (nealstewart) [inline lib/user_view.js]: Move this conditional up to the constructor instead. Just set `this._userTemplate = _.template(default)` if they don't pass a template function up there.

Comment (nealstewart) [inline lib/user_view.js]: This is nicer because the conditional happens immediately, rather than on every render.

**Comment by nealstewart** [comment]
LGTM pending :white_check_mark: and testing Docs + build on staging.

**Comment by nealstewart** [comment]
LGTM

File: test/index.js [hunk 4]
```diff
@@ -245,6 +256,27 @@ describe('User-List Component', function() {
...
+      userList = new UserList({
+        room: fakeRoom,
+        userTemplate: CUSTOM_TEMPLATE
+      });
+
+      userList.initialize(function(err) {
+        assert.ifError(err);
+
+        sinon.assert.calledWith(
+          UserList._UserView,
```
Comment (nealstewart) [inline test/index.js:271]: Are there no tests for user view?

Ideally there would be a test that makes the template a stub and ensures it was used inside of the UserView.

**Comment by colinmacdonald** [comment]
I feel like this option is rather dangerous to use. If the developer changed the structure of the user template at all it could break things. Also, if we add features to the user-list that affect the structure of the user template (or a custom template) it could also break the user-list.

**Comment by tmyrden** [comment]
Agreed that it is dangerous in certain situations. Especially if you're using it in conjunction with latest. But in certain situations, like the ExactTarget demo, you're forced to either fork the widget or use some really hacky jQuery to get in features you want, such as simply changing the user's name to act as a link. Obviously, the hope would be that the developer wouldn't just use this and blindly deploy it though. I think it would only see use in situations where the developer identified a strong case for changing the user view. Not adding a feature because they could break something during development doesn't seem like a good reason to not add it to me.

**Comment by colinmacdonald** [comment]
I'm not concerned about a developer breaking something during development, more about us adding a feature that could potentially break their user list if they are using a custom template. I guess what i would prefer is a more controlled way to customize the user view rather than passing an entire template. Not sure how this could be done without making it overly limited though.

**Comment by colinmacdonald** [comment]
LGTM

**Comment by nealstewart** [comment]
@tmyrden what happens if they change a name? does it re-render their user view?

**Comment by tmyrden** [comment]
@nealstewart Yea, the _handleUserMeta function handles changes to the userCache and renders the view new when things change. That's existing functionality.

**Comment by nealstewart** [comment]
Cool. Add the UserView unit test, and I think we're good.

File: test/user_view.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      };
+    });
+
+    it('renders the default template', function(done) {
+      var userView = new UserView(mockUserList);
+      sandbox.stub(userView, '_addLocalUserElement');
+      var templateSpy = sandbox.spy(userView, '_userTemplate');
+      userView.render({}, function() {
+        assert.equal(
+          templateSpy.returnValues[0].substr(0, 10),
```
Comment (colinmacdonald) [inline test/user_view.js]: This test is a bit weak where it only checks the first 10 characters of the template vs the default. Could just stub `_.template` and use `.calledWith()` to make sure the correct template is being used.

File: test/user_view.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        },
+        _avatars: {},
+        _enableUserOptions: true
+      };
+    });
+
+    it('renders the default template', function(done) {
+      var userView = new UserView(mockUserList);
+      sandbox.stub(userView, '_addLocalUserElement');
+      var templateSpy = sandbox.spy(userView, '_userTemplate');
```
Comment (colinmacdonald) [inline test/user_view.js]: spacing between vars and `#render`

File: test/user_view.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        );
+        done();
+      });
+    });
+
+    it('renders a custom template', function(done) {
+      var CUSTOM_TEMPLATE = '<div></div>';
+      mockUserList._userTemplate = sandbox.spy(_.template(CUSTOM_TEMPLATE));
+      var userView = new UserView(mockUserList);
+      sandbox.stub(userView, '_addLocalUserElement');
```
Comment (colinmacdonald) [inline test/user_view.js:59]: spacing

**Comment by colinmacdonald** [comment]
LGTM

**Pull request merged**